### PR TITLE
feat(goals): edge model for goal↔milestone links (PR 1 — foundation + migration)

### PR DESCRIPTION
**Pull request by gloriafolaron**
First of a phased refactor: the goal→milestone link is a single `zp_canvas_items.milestoneId` varchar column (single-select) — this moves it to **many-to-many edges** on the core `zp_entity_relationship` graph so a goal can be tracked by any number of milestones. Approved by Marcel; his clarification: **goal progress stays metric-defined** — milestones just render as a *list* below the goal (today capped at one).

## This PR (foundation only — no reader/writer changes, fully backward compatible)
- **`EntityRelationshipEnum::TrackedBy`** (`tracked_by`). Direction: `entityA = goal (GoalItem)`, `entityB = milestone (Ticket)`. New case rather than reusing `GeneratedFrom`/`MapsTo` (StrategyPro's wizard filters on those — reuse would hide goals from its candidate lists).
- **`update_sql_30524`** — backfills every `box='goal'` row's `milestoneId` into a `tracked_by` edge. Query-builder based (portable across MySQL/Postgres/MSSQL), idempotent, junk-guarded (skips empty/`'0'`/non-numeric and edges to deleted milestones). The `milestoneId` column is **kept**; a later migration drops it once readers are ported.
- Bump `dbVersion` 3.5.23 → 3.5.24.

## Verification
Live-DB run of the migration: 5/5 — creates the exact edge, source-tags it (`meta.source=milestoneId_migration`), idempotent on re-run, no dupes. `php -l` + Pint clean.

## Phasing (separate PRs)
2. cut readers (`getGoalsByMilestone`, reports, Tickets bridge, MCP) + writers to edges
3. multi-select UI
4. drop the `milestoneId` column

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by marcelfolaron** [comment]
**Status: triaged · Priority: P2 (foundation for the multi-milestone customer ask, #2939) · Next action: confirm migration idempotency test + cross-DB backfill · Owner: @marcelfolaron / @broskees**

*Reviewed at head `30e16a1` (incl. the Copilot-fix commit). PR 1 of the #3686→#3687→#3688 stack; base is `master`.*

**1. Intent:** Foundation-only, backward-compatible first step of the goal↔milestone refactor — adds `EntityRelationshipEnum::TrackedBy` (`tracked_by`) and `update_sql_30524` to backfill each `box='goal'` row's `milestoneId` into a `tracked_by` edge on the core `zp_entity_relationship` graph; the legacy column is kept, no readers/writers change yet.

**2. Change requests:**
1. **⚠ CR #1 (blocker) — data migration ships without an automated test.** `update_sql_30524` is load-bearing (a mis-backfill silently corrupts goal↔milestone links for every existing install) and it was *already* nearly shipped broken — the Copilot-fix commit reveals it was defined but **unregistered in `$dbUpdates`**, which would have silently skipped the backfill entirely. That near-miss is exactly why this needs a migration test, not just live-DB spot checks: assert the edge is created with the correct direction (`entityA=goal`, `entityB=milestone`), junk values (`''`/`'0'`/non-numeric/whitespace) are skipped, edges to deleted milestones are skipped, NULL author on unknown, and a second run is a no-op (idempotent).
2. **◐ CR #2 — confirm cross-DB portability of the query-builder backfill.** The PR claims MySQL/Postgres/MSSQL portability but verification was MySQL-only ("live DB"). `chunkById` + the id-scoped edge lookups should be portable, but confirm the backfill runs clean on at least Postgres before merge, since #106 (docker pgsql) means Postgres installs are a supported target.
3. **◐ CR #3 — `dbVersion` bump coordination across the stack.** This bumps 3.5.23 → 3.5.24 and #3687 adds `update_sql_30525` (→ 3.5.25). Confirm the three PRs merge in order (#3686 → #3687 → #3688) so the dbVersion sequence lands monotonically and the PR1→PR2-window top-up (30525) actually fires after 30524.
4. **✓ Enum choice is right.** New `TrackedBy` case rather than reusing `GeneratedFrom`/`MapsTo` correctly avoids StrategyPro's wizard filters hiding goals from candidate lists — good call, worth a code comment (already added per the fix commit).

**3. Acceptance checklist (must pass before merge):**
1. [ ] `update_sql_30524` migration test: correct edge direction, junk/deleted-milestone guards, NULL-author, idempotent re-run — committed and green.
2. [ ] Backfill verified on Postgres (not just MySQL); `$dbUpdates` registration confirmed (regression guard for the near-miss).
3. [ ] CI green (Pint + PHPStan L5 + `php -l` + full unit suite); stack merges #3686 → #3687 → #3688 in order.

**4. CI status:** ⚠️ Couldn't read check-runs this cycle — confirm in the UI. **Blocker flag:** a schema/data migration with no automated test (CR #1), made higher-risk by the just-fixed unregistered-migration bug. No behavior change to app code here, so runtime risk is low; migration correctness is the whole risk surface.

*Automated maintainer review — advisory only, not an approval, not a merge. Merge authority stays with @marcelfolaron / @broskees.*

File: app/Domain/Install/Repositories/Install.php [hunk 1]
```diff
@@ -2975,4 +2977,128 @@ public function update_sql_30523(): bool|array
...
+     * are not duplicated on re-run.
+     */
+    public function update_sql_30524(): bool|array
+    {
+        try {
+            if (! Schema::hasTable('zp_canvas_items')
+                || ! Schema::hasTable('zp_entity_relationship')
+                || ! Schema::hasTable('zp_tickets')) {
+                return true;
+            }
```
[the remark this answers is not in the document]
Comment (gloriafolaron) [reply app/Domain/Install/Repositories/Install.php]: Fixed in the latest push: the guard now resolves `$this->connection->getSchemaBuilder()` and calls hasTable() on it, so the existence check runs on the same connection as the migration queries (install_temp/target install connection included) instead of the default connection via the Schema facade.

File: app/Domain/Install/Repositories/Install.php [hunk 2]
```diff
@@ -2975,4 +2977,133 @@ public function update_sql_30523(): bool|array
...
+            // scan), then batch-inserts. Uses the installer's selected
+            // $this->connection (not the global DB facade), which may point at
+            // a temp/target install connection.
+            $this->connection->table('zp_canvas_items')
+                ->where('box', 'goal')
+                ->whereNotNull('milestoneId')
+                ->where('milestoneId', '<>', '')
+                ->where('milestoneId', '<>', '0')
+                ->orderBy('id')
+                ->chunkById(500, function ($goals) use ($now): void {
```
[the remark this answers is not in the document]
Comment (gloriafolaron) [reply app/Domain/Install/Repositories/Install.php:3038]: Fixed: the chunked backfill now selects only id, milestoneId, author (id is required for chunkById) — the heavy description/data text columns are never transferred.

File: app/Domain/Install/Repositories/Install.php [hunk 3]
```diff
@@ -2975,4 +2977,133 @@ public function update_sql_30523(): bool|array
...
+            // Guard on the installer's own connection (not the global Schema
+            // facade, which checks the default connection) so the existence
+            // check matches the connection the migration queries actually run
+            // against (may be a temp/target install connection).
+            $schema = $this->connection->getSchemaBuilder();
+            if (! $schema->hasTable('zp_canvas_items')
+                || ! $schema->hasTable('zp_entity_relationship')
+                || ! $schema->hasTable('zp_tickets')) {
+                return true;
+            }
```
[the remark this answers is not in the document]
Comment (gloriafolaron) [reply app/Domain/Install/Repositories/Install.php:3017]: Fixed: the guard now also checks $schema->hasColumn('zp_canvas_items', 'milestoneId') on the installer connection, so an install missing that column no-ops instead of throwing.

File: app/Domain/Install/Repositories/Install.php [hunk 4]
```diff
@@ -2975,4 +2977,137 @@ public function update_sql_30523(): bool|array
...
+                        $this->connection->table('zp_entity_relationship')
+                            ->where('relationship', EntityRelationshipEnum::TrackedBy->value)
+                            ->where('entityAType', 'GoalItem')
+                            ->where('entityBType', 'Ticket')
+                            ->whereIn('entityA', $goalIds)
+                            ->select('entityA', 'entityB')
+                            ->get() as $e
+                    ) {
+                        $existingEdges[(int) $e->entityA.':'.(int) $e->entityB] = true;
+                    }
```
[the remark this answers is not in the document]
Comment (gloriafolaron) [reply app/Domain/Install/Repositories/Install.php:3083]: Fixed in 3290f82ca — parenthesized the (int) casts in the existing-edge dedup key so the cast-vs-concat precedence is explicit, and the live-ticket check now filters type='milestone' so the backfill only creates edges to actual milestones.

File: app/Core/Support/EntityRelationshipEnum.php [hunk 5]
```diff
@@ -23,4 +23,13 @@ enum EntityRelationshipEnum: string
...
      */
     case MapsTo = 'maps_to';
+
+    /**
+     * Represents a "tracked by" relationship: a goal is tracked/measured by a
+     * milestone. Direction convention: entityA = goal (GoalItem),
+     * entityB = milestone (Ticket). The edge-based successor to the legacy
+     * single zp_canvas_items.milestoneId column (which is retained during the
+     * transition) so a goal can hold many milestones.
+     */
```
[the remark this answers is not in the document]
Comment (gloriafolaron) [reply app/Core/Support/EntityRelationshipEnum.php:35]: Fixed — reworded the TrackedBy docblock: the edge is informational (milestones render as a list on the goal) and does NOT drive goal progress, which stays metric-defined. Matches the product rule that milestones are informational only.

**Comment by marcelfolaron** [comment]
**Review (advisory / re-review) — @gloriafolaron** · reviewed at HEAD `969f87c` (new commits since the 07-24 pass).

**Intent:** PR 1 of the goal↔milestone refactor — add `EntityRelationshipEnum::TrackedBy` and migration `update_sql_30524` backfilling every `box='goal'` row's legacy `zp_canvas_items.milestoneId` into a `tracked_by` edge on `zp_entity_relationship` (portable, idempotent, junk-guarded); dbVersion 3.5.23→3.5.24. Foundation only, column retained.

**Since my last review — prior robustness asks all addressed** (confirmed in commits): migration registered in `$dbUpdates`; chunked backfill selecting only id/milestoneId/author; UTC `gmdate`; installer `$this->connection` + per-connection schema builder; `milestoneId` column existence guard; live-ticket filter `type='milestone'` AND `status<>-1`; `sprintf('%d:%d', …)` explicit edge-key; NULL author for unknown. Good progress.

**Change requests / confirm before merge:**
1. **CI blocker — `phpstan` FAILING at HEAD** (3 annotations). Level-5 is a merge gate; resolve before merge. See CR-CI.
2. **Stack merge-order** — base is `master`, but the diff-vs-master carries the whole lineage. Confirm the order **#3686 → #3687 → #3688 → #3690/#3691**, retargeting each as its parent lands, so `update_sql_30524`/`30525` apply once and in sequence and dbVersion lands cleanly. Release note must call out the dbVersion bump.
3. **Backfill idempotency at scale** — `zp_entity_relationship` has no composite unique on (tracked_by, goal, milestone). Confirm a re-run of 30524 on a populated instance creates **0** new edges (the chunked existing-edge dedup fully covers the scan).

**Acceptance checklist (must pass before merge):**
- [ ] PHPStan level 5 green (**currently failing**) · Pint · full unit suite green.
- [ ] Migration verified idempotent on a populated DB (re-run = 0 new edges); dbVersion = 3.5.24 exactly once, in stack order.
- [ ] Release/migration note documents the dbVersion bump + one-time backfill.

**CI status @ `969f87c`:** 🔴 **`phpstan` FAILING (3) — high priority, merge gate.** `NPM Security Audit` red is repo-wide/pre-existing (red on base too). pint · unit · acceptance · build · CodeQL green.

Advisory only — not approving/merging. Merge authority stays with @marcelfolaron / @broskees.

**Comment by marcelfolaron** [comment]
**Review (advisory / re-review) — @gloriafolaron** · reviewed at HEAD `e54a3513` (one new commit since my `969f87c` pass). PR 1 of the goal↔milestone stack.

**1. Intent (unchanged):** Foundation-only — `EntityRelationshipEnum::TrackedBy` + `update_sql_30524` backfilling each `box='goal'` row's `milestoneId` into a `tracked_by` edge; dbVersion 3.5.23→3.5.24; legacy column kept.

**What changed since `969f87c`:** a single commit `e54a3513` — narrows the `getSchemaBuilder()` call on `$this->connection` (typed `ConnectionInterface`, which lacks the method) with an inline `@var` to the concrete `Connection` so **PHPStan level 5 passes**. No behavior change. This directly targets my carried CR #1 (the phpstan-L5 merge gate) — good.

**2. Change requests:**
1. **◐ CR #1 (carried, now likely clearing) — confirm phpstan L5 is actually green on `e54a3513`.** The `@var` narrowing should resolve the 3 annotations I flagged at `969f87c`; confirm the `phpstan` job is green on this exact head rather than assuming the inline hint satisfied it.
2. **⚠ CR #2 (carried, the real gap) — the migration still has no automated test.** `update_sql_30524` is load-bearing (a mis-backfill corrupts goal↔milestone links for every install) and already had one near-miss (unregistered in `$dbUpdates`). Add a migration test: correct edge direction (`entityA=goal`, `entityB=milestone`), junk/`'0'`/non-numeric skipped, deleted-milestone (`status=-1`) skipped, NULL author on unknown, idempotent re-run = 0 new edges. Live-DB spot checks ≠ a committed regression test.
3. **◐ CR #3 (carried) — cross-DB + stack order.** Confirm the query-builder backfill runs clean on Postgres (verification has been MySQL-only), and that the stack merges **#3686 → #3687 → #3688** so `update_sql_30524/30525` apply once, in order, and dbVersion lands monotonically. Release note must call out the dbVersion bump.

**3. Acceptance checklist (must pass before merge):**
- [ ] `update_sql_30524` migration test (direction, junk/deleted guards, NULL author, idempotent) committed and green (CR #2).
- [ ] Backfill verified on Postgres (CR #3); stack merges #3686 → #3687 → #3688 in order; dbVersion bump in release notes.
- [ ] CI green on `e54a3513`: Pint + PHPStan L5 (the fix target) + full unit suite.

**4. CI status @ `e54a3513`:** ⚠️ couldn't read check-runs this cycle — confirm in the UI. The new commit is a PHPStan-L5 type narrowing that should clear the `phpstan` gate that was red at `969f87c`; `NPM Security Audit` red is repo-wide/pre-existing. Migration-correctness (no test) remains the whole risk surface here since there's no app-behavior change.

Advisory only — not approving/merging. Merge authority stays with @marcelfolaron / @broskees.
